File: tools/asr/vosk-asr-service.mjs
import { createRequire } from 'node:module';
import http from 'node:http';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const require = createRequire(import.meta.url);
const DEFAULT_PORT = 3301;
const MAX_JSON_BYTES = 48 * 1024 * 1024;
const DEFAULT_SAMPLE_RATE = 16000;

const args = parseArgs(process.argv.slice(2));
const scriptDir = path.dirname(fileURLToPath(import.meta.url));
const repoRoot = path.resolve(args['repo-root'] || process.env.YACHIYO_REPO_ROOT || path.join(scriptDir, '..', '..'));
const port = clampInt(args.port, DEFAULT_PORT, 1, 65535);
const modelCache = new Map();

let voskModule = null;

function parseArgs(argv) {
  const result = {};
  for (let i = 0; i < argv.length; i++) {
    const value = argv[i] || '';
    if (!value.startsWith('--')) continue;
    const eq = value.indexOf('=');
    if (eq >= 0) {
      result[value.slice(2, eq)] = value.slice(eq + 1);
      continue;
    }
    const key = value.slice(2);
    result[key] = argv[i + 1] && !argv[i + 1].startsWith('--') ? argv[++i] : 'true';
  }
  return result;
}

function clampInt(value, fallback, min, max) {
  const numeric = Number(value);
  if (!Number.isFinite(numeric)) return fallback;
  return Math.min(Math.max(Math.round(numeric), min), max);
}

function jsonResponse(response, statusCode, body) {
  const payload = Buffer.from(JSON.stringify(body), 'utf8');
  response.writeHead(statusCode, {
    'Content-Type': 'application/json; charset=utf-8',
    'Content-Length': payload.length,
    'Cache-Control': 'no-store',
    Connection: 'close'
  });
  response.end(payload);
}

function readRequestBody(request) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    let total = 0;
    request.on('data', (chunk) => {
      total += chunk.length;
      if (total > MAX_JSON_BYTES) {
        reject(new Error('ASR request is too large.'));
        request.destroy();
        return;
      }
      chunks.push(chunk);
    });
    request.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
    request.on('error', reject);
  });
}

function loadVosk() {
  if (voskModule) return voskModule;
  try {
    voskModule = require('vosk');
    voskModule.setLogLevel?.(-1);
    return voskModule;
  } catch (error) {
    throw new Error(`Vosk package is not installed. Run npm install in the project root. ${error.message}`);
  }
}

function resolveModelPath(modelPath) {
  const raw = String(modelPath || '').trim();
  const candidate = raw
    ? (path.isAbsolute(raw) ? raw : path.resolve(repoRoot, raw))
    : path.resolve(repoRoot, 'models', 'vosk', 'vosk-model-small-cn-0.22');
  if (!fs.existsSync(candidate) || !fs.statSync(candidate).isDirectory()) {
    throw new Error(`Vosk model directory not found: ${candidate}. Run npm run install:vosk-model, or set Settings -> ASR -> Model Path to an installed Vosk model directory.`);
  }
  return candidate;
}

function getModel(modelPath) {
  const resolved = resolveModelPath(modelPath);
  if (modelCache.has(resolved)) return modelCache.get(resolved);
  const vosk = loadVosk();
  const model = new vosk.Model(resolved);
  modelCache.set(resolved, model);
  return model;
}

function readAscii(buffer, offset, length) {
  return buffer.slice(offset, offset + length).toString('ascii');
}

function parseWav(buffer) {
  if (buffer.length < 44 || readAscii(buffer, 0, 4) !== 'RIFF' || readAscii(buffer, 8, 4) !== 'WAVE') {
    return null;
  }

  let offset = 12;
  let audioFormat = 0;
  let channels = 0;
  let sampleRate = DEFAULT_SAMPLE_RATE;
  let bitsPerSample = 0;
  let dataStart = -1;
  let dataSize = 0;

  while (offset + 8 <= buffer.length) {
    const chunkId = readAscii(buffer, offset, 4);
    const chunkSize = buffer.readUInt32LE(offset + 4);
    offset += 8;
    if (chunkId === 'fmt ') {
      audioFormat = buffer.readUInt16LE(offset);
      channels = buffer.readUInt16LE(offset + 2);
      sampleRate = buffer.readUInt32LE(offset + 4);
      bitsPerSample = buffer.readUInt16LE(offset + 14);
    } else if (chunkId === 'data') {
      dataStart = offset;
      dataSize = Math.min(chunkSize, buffer.length - offset);
      break;
    }
    offset += chunkSize + (chunkSize % 2);
  }

  if (audioFormat !== 1 || channels !== 1 || bitsPerSample !== 16 || dataStart < 0) {
    throw new Error('Vosk ASR expects mono 16-bit PCM WAV audio.');
  }

  return {
    sampleRate,
    pcm: buffer.slice(dataStart, dataStart + dataSize)
  };
}

function normalizeVoskResult(value) {
  if (!value) return {};
  if (typeof value === 'string') {
    try {
      return JSON.parse(value);
    } catch (_) {
      return { text: value };
    }
  }
  return value;
}

async function recognize(input) {
  const audioBase64 = String(input.audioBase64 || '').trim();
  if (!audioBase64) throw new Error('audioBase64 is required.');
  const model = getModel(input.modelPath);
  const audio = Buffer.from(audioBase64, 'base64');
  const wav = parseWav(audio);
  const sampleRate = clampInt(wav?.sampleRate || input.sampleRate, DEFAULT_SAMPLE_RATE, 8000, 48000);
  const pcm = wav?.pcm || audio;
  if (!pcm.length) throw new Error('Audio payload is empty.');

  const vosk = loadVosk();
  const recognizer = new vosk.Recognizer({ model, sampleRate });
  try {
    if (input.words && typeof recognizer.setWords === 'function') {
      recognizer.setWords(true);
    }
    const maxAlternatives = clampInt(input.maxAlternatives, 0, 0, 10);
    if (maxAlternatives > 0 && typeof recognizer.setMaxAlternatives === 'function') {
      recognizer.setMaxAlternatives(maxAlternatives);
    }
    recognizer.acceptWaveform(pcm);
    const result = normalizeVoskResult(recognizer.finalResult());
    const alternatives = Array.isArray(result.alternatives)
      ? result.alternatives
          .filter((item) => String(item?.text || '').trim())
          .sort((left, right) => Number(right?.confidence) - Number(left?.confidence))
      : [];
    return {
      text: String(alternatives[0]?.text || result.text || '').trim(),
      result,
      sampleRate,
      modelPath: resolveModelPath(input.modelPath)
    };
  } finally {
    recognizer.free?.();
  }
}

const server = http.createServer(async (request, response) => {
  const url = new URL(request.url || '/', `http://${request.headers.host || '127.0.0.1'}`);
  if (request.method === 'GET' && url.pathname === '/healthz') {
    jsonResponse(response, 200, { ok: true, service: 'vosk-asr' });
    return;
  }
  if (request.method !== 'POST' || url.pathname !== '/api/asr') {
    jsonResponse(response, 404, { success: false, error: 'Not Found' });
    return;
  }

  try {
    const rawBody = await readRequestBody(request);
    const input = rawBody ? JSON.parse(rawBody) : {};
    const data = await recognize(input);
    jsonResponse(response, 200, { success: true, data });
  } catch (error) {
    jsonResponse(response, 500, { success: false, error: error.message || 'ASR failed' });
  }
});

server.listen(port, '127.0.0.1', () => {
  process.stdout.write(`Yachiyo Vosk ASR service listening on http://127.0.0.1:${port}\n`);
});

function shutdown() {
  for (const model of modelCache.values()) {
    model.free?.();
  }
  modelCache.clear();
  server.close(() => process.exit(0));
}

process.on('SIGINT', shutdown);
process.on('SIGTERM', shutdown);

File: tools/asr/vosk_asr_service.py
import argparse
import base64
import json
import os
import wave
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from io import BytesIO


DEFAULT_PORT = 3301
DEFAULT_SAMPLE_RATE = 16000
MAX_JSON_BYTES = 48 * 1024 * 1024
MODEL_CACHE = {}


def clamp_int(value, fallback, min_value, max_value):
    try:
        numeric = int(round(float(value)))
    except (TypeError, ValueError):
        return fallback
    return min(max(numeric, min_value), max_value)


def resolve_model_path(repo_root, model_path):
    raw = str(model_path or "").strip()
    if raw:
        candidate = raw if os.path.isabs(raw) else os.path.abspath(os.path.join(repo_root, raw))
    else:
        candidate = os.path.abspath(os.path.join(repo_root, "models", "vosk", "vosk-model-small-cn-0.22"))
    if not os.path.isdir(candidate):
        raise RuntimeError(
            "Vosk model directory not found: "
            + candidate
            + ". Run npm run install:vosk-model, or set Settings -> ASR -> Model Path to an installed Vosk model directory."
        )
    return candidate


def get_model(repo_root, model_path):
    resolved = resolve_model_path(repo_root, model_path)
    if resolved in MODEL_CACHE:
        return MODEL_CACHE[resolved], resolved

    from vosk import Model, SetLogLevel

    SetLogLevel(-1)
    model = Model(resolved)
    MODEL_CACHE[resolved] = model
    return model, resolved


def parse_audio(audio_bytes, sample_rate):
    try:
        with wave.open(BytesIO(audio_bytes), "rb") as wav:
            if wav.getnchannels() != 1 or wav.getsampwidth() != 2:
                raise RuntimeError("Vosk ASR expects mono 16-bit PCM WAV audio.")
            return wav.readframes(wav.getnframes()), wav.getframerate()
    except wave.Error:
        return audio_bytes, sample_rate


def recognize(repo_root, payload):
    audio_base64 = str(payload.get("audioBase64") or "").strip()
    if not audio_base64:
        raise RuntimeError("audioBase64 is required.")

    sample_rate = clamp_int(payload.get("sampleRate"), DEFAULT_SAMPLE_RATE, 8000, 48000)
    model, model_path = get_model(repo_root, payload.get("modelPath"))
    audio_bytes = base64.b64decode(audio_base64)
    pcm, wav_sample_rate = parse_audio(audio_bytes, sample_rate)
    sample_rate = clamp_int(wav_sample_rate, sample_rate, 8000, 48000)
    if not pcm:
        raise RuntimeError("Audio payload is empty.")

    from vosk import KaldiRecognizer

    recognizer = KaldiRecognizer(model, sample_rate)
    if payload.get("words"):
        recognizer.SetWords(True)
    max_alternatives = clamp_int(payload.get("maxAlternatives"), 0, 0, 10)
    if max_alternatives > 0:
        recognizer.SetMaxAlternatives(max_alternatives)
    recognizer.AcceptWaveform(pcm)
    result = json.loads(recognizer.FinalResult() or "{}")
    alternatives = sorted(
        (
            item
            for item in result.get("alternatives", [])
            if isinstance(item, dict) and str(item.get("text") or "").strip()
        ),
        key=lambda item: float(item.get("confidence") or 0),
        reverse=True,
    )
    return {
        "text": str((alternatives[0].get("text") if alternatives else result.get("text")) or "").strip(),
        "result": result,
        "sampleRate": sample_rate,
        "modelPath": model_path,
    }


class AsrHandler(BaseHTTPRequestHandler):
    repo_root = os.getcwd()

    def log_message(self, format, *args):
        return

    def send_json(self, status, body):
        encoded = json.dumps(body, ensure_ascii=False).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(encoded)))
        self.send_header("Cache-Control", "no-store")
        self.send_header("Connection", "close")
        self.end_headers()
        self.wfile.write(encoded)

    def do_GET(self):
        if self.path.split("?", 1)[0] == "/healthz":
            self.send_json(200, {"ok": True, "service": "vosk-asr-python"})
            return
        self.send_json(404, {"success": False, "error": "Not Found"})

    def do_POST(self):
        if self.path.split("?", 1)[0] != "/api/asr":
            self.send_json(404, {"success": False, "error": "Not Found"})
            return

        try:
            content_length = clamp_int(self.headers.get("Content-Length"), 0, 0, MAX_JSON_BYTES + 1)
            if content_length > MAX_JSON_BYTES:
                raise RuntimeError("ASR request is too large.")
            raw_body = self.rfile.read(content_length).decode("utf-8")
            payload = json.loads(raw_body or "{}")
            data = recognize(self.repo_root, payload)
            self.send_json(200, {"success": True, "data": data})
        except Exception as error:
            self.send_json(500, {"success": False, "error": str(error) or "ASR failed"})


def main():
    parser = argparse.ArgumentParser(description="Yachiyo local Vosk ASR service.")
    parser.add_argument("--port", type=int, default=DEFAULT_PORT)
    parser.add_argument("--repo-root", default=os.environ.get("YACHIYO_REPO_ROOT") or os.getcwd())
    args = parser.parse_args()

    AsrHandler.repo_root = os.path.abspath(args.repo_root)
    server = ThreadingHTTPServer(("127.0.0.1", clamp_int(args.port, DEFAULT_PORT, 1, 65535)), AsrHandler)
    print(f"Yachiyo Python Vosk ASR service listening on http://127.0.0.1:{args.port}", flush=True)
    server.serve_forever()


if __name__ == "__main__":
    main()
